plugins/hlint-hls-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,373 @@
+instance Binary GetHlintSettings
+
+type instance RuleResult GetHlintSettings = (ParseFlags, [Classify], Hint)
+
+getHlintSettingsRule :: HlintUsage -> Rules ()
+getHlintSettingsRule usage =
+    defineNoFile $ \GetHlintSettings ->
+      liftIO $ case usage of
+          HlintEnabled cmdArgs -> argsSettings cmdArgs
+          HlintDisabled -> fail "hlint configuration unspecified"
+
+-- ---------------------------------------------------------------------
+
+codeActionProvider :: CodeActionProvider
+codeActionProvider _lf ideState plId docId _ context = Right . LSP.List . map CACodeAction <$> getCodeActions
+  where
+
+    getCodeActions = do
+        applyOne <- applyOneActions
+        diags <- getDiagnostics ideState
+        let docNfp = toNormalizedFilePath' <$> uriToFilePath' (docId ^. LSP.uri)
+            numHintsInDoc = length
+              [d | (nfp, _, d) <- diags
+                 , d ^. LSP.source == Just "hlint"
+                 , Just nfp == docNfp
+              ]
+        -- We only want to show the applyAll code action if there is more than 1
+        -- hint in the current document
+        if numHintsInDoc >= 2 then do
+          applyAll <- applyAllAction
+          pure $ applyAll:applyOne
+        else
+          pure applyOne
+
+    applyAllAction = do
+      let args = Just [toJSON (docId ^. LSP.uri)]
+      cmd <- mkLspCommand plId "applyAll" "Apply all hints" args
+      pure $ LSP.CodeAction "Apply all hints" (Just LSP.CodeActionQuickFix) Nothing Nothing (Just cmd)
+
+    applyOneActions :: IO [LSP.CodeAction]
+    applyOneActions = catMaybes <$> mapM mkHlintAction (filter validCommand diags)
+
+    -- |Some hints do not have an associated refactoring
+    validCommand (LSP.Diagnostic _ _ (Just (LSP.StringValue code)) (Just "hlint") _ _ _) =
+      code /= "Eta reduce"
+    validCommand _ = False
+
+    LSP.List diags = context ^. LSP.diagnostics
+
+    mkHlintAction :: LSP.Diagnostic -> IO (Maybe LSP.CodeAction)
+    mkHlintAction diag@(LSP.Diagnostic (LSP.Range start _) _s (Just (LSP.StringValue code)) (Just "hlint") m _ _) =
+      Just . codeAction <$> mkLspCommand plId "applyOne" title (Just args)
+     where
+       codeAction cmd = LSP.CodeAction title (Just LSP.CodeActionQuickFix) (Just (LSP.List [diag])) Nothing (Just cmd)
+       title = "Apply hint:" <> head (T.lines m)
+       -- need 'file', 'start_pos' and hint title (to distinguish between alternative suggestions at the same location)
+       args = [toJSON (AOP (docId ^. LSP.uri) start code)]
+    mkHlintAction (LSP.Diagnostic _r _s _c _source _m _ _) = return Nothing
+
+-- ---------------------------------------------------------------------
+
+applyAllCmd :: CommandFunction Uri
+applyAllCmd lf ide uri = do
+  let file = maybe (error $ show uri ++ " is not a file.")
+                    toNormalizedFilePath'
+                   (uriToFilePath' uri)
+  withIndefiniteProgress lf "Applying all hints" Cancellable $ do
```

Comment:
I haven't had a chance to try out this change yet but is applying the hints particularly slow? #381

Reply:
Well, in my experienc appliying one is fast in the general case but sometimes you hit a hiccup. Applying all of them if the file is large and has many hints can be slow enough in the general case to deserve the progress alert.
Ideally the notification should depend on the time spent, as noted in the linked issue
A `withIndefiniteProgressWithThreshold` could be handy.

Reply:
Maybe we could do it only when applying all, while there is no threshold